runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,346 @@
+          if in_brackets and (char ~= "]" or bracket_val == "") then
+            local res
+            if char == "-" then
+              res = char
+            elseif bracket_val ~= "" and char == "!" then
+              res = "^"
+            elseif char == "**" then
+              res = ""
+            else
+              res = escape(char)
+            end
+            bracket_val = bracket_val..res
+          else
+            if char == "{" then
+              in_braces = true
+            elseif char == "[" then
+              in_brackets = true
+            elseif char == "}" then
+              local choices = split(brace_val, ",")
+              local parsed_choices = {}
+              for _, choice in ipairs(choices) do
+                table.insert(parsed_choices, parse(choice))
+              end
+              append(vim.tbl_flatten(parsed_choices))
+              in_braces = false
+              brace_val = ""
+            elseif char == "]" then
+              append({"["..bracket_val.."]"})
+              in_brackets = false
+              bracket_val = ""
+            elseif char == "?" then
+              append({non_path_sep})
+            elseif char == "*" then
+              append({non_path_sep.."-"})
+            else
+              append({escape(char)})
+            end
+          end
+        end
+      end
+
+      if not last_seg and (segments[i+1] ~= "**" or i+1 < #segments) then
+        append({path_sep})
+      end
+    end
+  end
+
+  return patterns
+end
+
+---@private
+--- Implementation of LSP 3.17.0's pattern matching: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#pattern
+--- Modeled after VSCode's implementation: https://github.com/microsoft/vscode/blob/0319eed971719ad48e9093daba9d65a5013ec5ab/src/vs/base/common/glob.ts#L509
+---
+---@param pattern string|table The glob pattern (raw or parsed) to match.
+---@param s string The string to match against pattern.
+---@return bool Whether or not pattern matches s.
+local function match(pattern, s)
+  if type(pattern) == "string" then
+    pattern = parse(pattern)
+  end
+  -- Since Lua's built-in string pattern matching does not have an alternate
+  -- operator like '|', `parse` will construct one pattern for each possible
+  -- alternative. Any pattern that matches thus matches the glob.
+  for _, p in ipairs(pattern) do
+    if s:match("^"..p.."$") then return true end
+  end
+  return false
+end
+
+---@private
+--- Joins filepath elements by platform-specific separator.
+---
+---@param args ... The path elements. The first element must be absolute.
+---@return string The joined path.
+local function filepath_join(...)
+  local dir = ...
+  return table.concat({...}, dir:match('^([a-zA-Z]:)(.*)') and "\\" or "/")
+end
+
+-- Cache of libuv handles per directory, per LSP client, per registration ID.
+local watched_dirs = {}
+
+local change_queue = {}
+local change_cache = {}
+local queue_timers = {}
```

Comment:
These variables may be read from or written to concurrently by multiple async callbacks. Is there any explicit synchronization necessary for things like this?